src/dynamic-launcher.c

```diff
@@ -0,0 +1,794 @@
+  if (prefixed_exec_strv == NULL)
+    {
+      g_set_error (error,
+                   XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_FAILED,
+                   "DynamicLauncher install not supported for: %s", xdp_app_info_get_id (xdp_app_info));
+      return NULL;
+    }
+
+  prefixed_exec = g_strjoinv (" ", prefixed_exec_strv);
+  g_key_file_set_value (key_file, G_KEY_FILE_DESKTOP_GROUP, "Exec", prefixed_exec);
+
+  tryexec_path = xdp_app_info_get_tryexec_path (xdp_app_info);
+  if (tryexec_path == NULL)
+    g_debug ("Couldn't get tryexec path for app %s", xdp_app_info_get_id (xdp_app_info));
+  else
+    g_key_file_set_value (key_file, G_KEY_FILE_DESKTOP_GROUP, "TryExec", tryexec_path);
+
+  desktop_app_info = g_desktop_app_info_new_from_keyfile (key_file);
+  if (desktop_app_info == NULL)
+    {
+      g_set_error (error,
+                   XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Desktop entry given to Install() not valid");
+      return NULL;
+    }
+
+  /* Write the icon last so it's only on-disk if other checks passed */
+  {
+    g_autoptr(GIcon) icon = NULL;
+    GBytes *icon_bytes;
+    g_autoptr(GFile) icon_file = NULL;
+    g_autoptr(GFileOutputStream) icon_stream = NULL;
+    gssize bytes_written;
+
+    icon = g_icon_deserialize (icon_v);
+    g_assert (G_IS_BYTES_ICON (icon));
+    icon_bytes = g_bytes_icon_get_bytes (G_BYTES_ICON (icon));
+
+    g_mkdir_with_parents (icon_dir, 0755);
+    icon_file = g_file_new_for_path (icon_path);
+    icon_stream = g_file_replace (icon_file, NULL, FALSE, G_FILE_CREATE_REPLACE_DESTINATION, NULL, error);
+    if (icon_stream == NULL)
+      return NULL;
+    bytes_written = g_output_stream_write_bytes (G_OUTPUT_STREAM (icon_stream), icon_bytes, NULL, error);
+    if (bytes_written < 0)
+      return NULL;
+  }
+
+  if (out_icon_path)
+    *out_icon_path = g_steal_pointer (&icon_path);
+
+  return g_steal_pointer (&key_file);
+}
+
+static gboolean
+handle_install (XdpDynamicLauncher    *object,
+                GDBusMethodInvocation *invocation,
+                const gchar           *arg_token,
+                const gchar           *arg_desktop_file_id,
+                const gchar           *arg_desktop_entry,
+                GVariant              *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  const char *app_id = xdp_app_info_get_id (call->app_info);
+  g_autoptr(GVariant) launcher_data = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GDesktopAppInfo) app_info = NULL;
+  g_autoptr(GKeyFile) desktop_keyfile = NULL;
+  g_autofree char *icon_path = NULL;
+  g_autofree char *desktop_dir = NULL;
+  g_autofree char *desktop_path = NULL;
+  g_autofree char *link_path = NULL;
+  g_autofree char *relative_path = NULL;
+  g_autoptr(GFile) link_file = NULL;
+  gint result = -1;
+
+  launcher_data = get_launcher_data_and_revoke_token (arg_token);
```

Comment:
I think it would be useful to make `get_launcher_data_and_revoke_token()` validate `arg_token` with `g_uuid_string_is_valid()`